### DISCUSSION TITLE
Crazyflie Hover mode with zranger2 and a custom Positioning system

### DISCUSSION THREAD

**Discussion by valiokei** [bitcraze/discussions#1496] [Q&A]
Hello, I'm wondering how can i take off in **hover mode** using the cfclient when the drone is equipped with a zRanger2 and a custom positioning system.
From the console, I can see that both the systems are recognized and working (the drone state, x,y, and z are stable and not diverging before landing). 
Also,I've mounted a Flowdeck alongside my positioning system and I can take off.

Do you have any clue how to let the crazyflie take off using just these 2 decks?

**Comment by knmcguire** [comment]
Hi!

Some decks that provide full positioning automatically select a different state estimation. For the z-ranger that is the complementary filter (that's the default) as it doesn't provide x/y, for the flowdeck that is the [extended kalman filter](https://github.com/bitcraze/crazyflie-firmware/blob/7f1f78251fccdb355d672d579997238258e367d2/src/deck/drivers/src/flowdeck_v1v2.c#L251).

What I suspect 2 things that are happening:
- Your crazyflie still uses the complementary filter and doesn't have x and y in the state (as you noticed, those are set to zero), so you can adjust by assigning 2 to the stabilizer.estimator parameter: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#stabilizer
- The cfclient is hardcoded to only make some flight behaviors available once certain decks are attached. You'll probably need to change that in the code of the cfclient: https://github.com/bitcraze/crazyflie-clients-python/blob/5402e5db1937b2938f5af183670f7c2b82ed832c/src/cfclient/ui/tabs/FlightTab.py#L810-L826

**Reply by valiokei** [reply]
Hi, thanks for the kind and fast response.
I already setted the EKF for the estimator.
I'm not fixed on using the cfclient, also a python script could be a good choice.
Yesterday I was looking at this example script '[initial_position.py]( https://github.com/bitcraze/crazyflie-lib-python/blob/7feb51e26ff3fdcb8da0c3352326499f4b07ab7a/examples/positioning/initial_position.py)'.
But also in this example, I couldn't find a way to set the position hold flight mode properly. 

Is there a way in Python to hover using the position hold flight mode?
Basically, I want to be sure that my positioning system is ok for flight.

**Reply by knmcguire** [reply]
Yes absolutely. There are several example scripts you can try out here on the crazyflie python library: https://github.com/bitcraze/crazyflie-lib-python

Also there is a step by step guide for the motion commander, which does hover position hold and velocity in horizontal plane: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_motion_commander/
